File: cpp/control/pid_controller.cpp
#include <iostream>
#include <stdexcept>

class PIDController {
public:
    PIDController(double kp, double ki, double kd, double setpoint)
        : kp_(kp), ki_(ki), kd_(kd), setpoint_(setpoint) {}

    double update(double measurement, double dt) {
        if (dt <= 0.0) {
            throw std::invalid_argument("dt must be positive");
        }

        const double error = setpoint_ - measurement;
        integral_ += error * dt;
        const double derivative = (error - previous_error_) / dt;
        previous_error_ = error;
        return kp_ * error + ki_ * integral_ + kd_ * derivative;
    }

private:
    double kp_;
    double ki_;
    double kd_;
    double setpoint_;
    double integral_ = 0.0;
    double previous_error_ = 0.0;
};

int main() {
    PIDController pid(0.8, 0.04, 0.12, 10.0);
    double wheelSpeed = 0.0;

    for (int i = 0; i < 10; ++i) {
        const double command = pid.update(wheelSpeed, 0.1);
        wheelSpeed += command * 0.05;
        std::cout << "command=" << command << " wheelSpeed=" << wheelSpeed << '\n';
    }
}

File: python/control/pid_controller.py
"""Simple PID controller demo for robotics control loops."""

from dataclasses import dataclass


@dataclass
class PIDController:
    kp: float
    ki: float
    kd: float
    setpoint: float
    integral: float = 0.0
    previous_error: float = 0.0

    def update(self, measurement: float, dt: float) -> float:
        if dt <= 0:
            raise ValueError("dt must be positive")

        error = self.setpoint - measurement
        self.integral += error * dt
        derivative = (error - self.previous_error) / dt
        self.previous_error = error
        return self.kp * error + self.ki * self.integral + self.kd * derivative


if __name__ == "__main__":
    pid = PIDController(kp=0.9, ki=0.05, kd=0.1, setpoint=1.0)
    position = 0.0

    for step in range(1, 16):
        command = pid.update(position, dt=0.1)
        position += command * 0.08
        print(f"step={step:02d} command={command:+.3f} position={position:.3f}")
